test/functional/tests/ci/test_recovery.py

```diff
@@ -0,0 +1,209 @@
+
+    with TestRun.step("Start caches and add cores."):
+        caches, cores = [], []
+        for (cache_dev, core_dev) in zip(cache_devs, core_devs):
+            cache = casadm.start_cache(cache_dev, cache_mode=CacheMode.WB)
+            core = cache.add_core(core_dev)
+            caches.append(cache)
+            cores.append(core)
+
+    with TestRun.step("Create filesystem on core devices."):
+        filesystems = [Filesystem.xfs, Filesystem.ext3, Filesystem.ext4]
+        for (core, filesystem) in zip(cores, filesystems):
+            core.create_filesystem(filesystem)
+
+    with TestRun.step("Mount cache devices."):
+        for (cache, core) in zip(caches, cores):
+            core_mnt_point = f"{mount_point}-{cache.cache_id}-{core.core_id}"
+            core.mount(core_mnt_point)
+
+            with TestRun.step("Run IO"):
+                dd = (
+                    Dd()
+                    .input("/dev/urandom")
+                    .output(f"{core_mnt_point}/test")
+                    .count(1)
+                    .block_size(Size(50, Unit.MegaByte))
+                )
+                dd.run()
+                sync()
+
+    with TestRun.step("Calculate cache md5sums before unplug."):
+        core_mnt_md5s_before = [File(f"{core.mount_point}/test").md5sum() for core in cores]
+
+    with TestRun.step("Umount core devices."):
+        for core in cores:
+            core.unmount()
+
+    with TestRun.step("Dirty stop"):
+        dirty_stop(cache_disk, caches)
+
+    with TestRun.step("Load caches with metadata."):
+        for cache_dev in cache_devs:
+            casadm.load_cache(cache_dev)
+
+    with TestRun.step("Mount core devices."):
+        for core, cache in zip(cores, caches):
+            core_mnt_point = f"{mount_point}-{cache.cache_id}-{core.core_id}"
+            core.mount(core_mnt_point)
+            if not check_if_file_exists(f"{core_mnt_point}/test"):
+                TestRun.LOGGER.error(f"Mounting core device {core_mnt_point} failed.")
+
+    with TestRun.step("Calculate cache md5sums after recovery."):
+        core_mnt_md5s_after = [File(f"{core.mount_point}/test").md5sum() for core in cores]
+
+    with TestRun.step("Compare md5 sums for cores and core devices"):
+        if core_mnt_md5s_before != core_mnt_md5s_after:
+            TestRun.fail(
+                f"MD5 sums of core before and after does not match."
+                f"Expected: {core_mnt_md5s_before}, Actual: {core_mnt_md5s_after}"
+            )
+
+    with TestRun.step("Umount core devices."):
+        for core in cores:
+            core.unmount()
+
+
+def dirty_stop(cache_disk, caches: list):
+    with TestRun.step("Turn off cache devices."):
+        cache_disk.unplug()
+
+    with TestRun.step("Stop caches without flushing."):
+        for cache in caches:
+            with pytest.raises(CmdException) as ex:
+                cache.stop(no_data_flush=True)
+            assert "Failed to stop cache." in str(ex.value)
```

Comment:
You should try to keep off `asserts` and use `TestRun.fail()` instead. Or even better, you can try with `api/cas/cli_messages.py` api.